File: client/go/src/example/device/device.go
package device

import (
	"fmt"
	"context"
	"biostar/service/device"
	"google.golang.org/grpc"
)

type DeviceSvc struct {
	client device.DeviceClient
}

func NewDeviceSvc(conn *grpc.ClientConn) *DeviceSvc {
	return &DeviceSvc{
		client: device.NewDeviceClient(conn),
	}
}

func (s *DeviceSvc) GetInfo(deviceID uint32) (*device.FactoryInfo, error) {
	req := &device.GetInfoRequest{
		DeviceID: deviceID,
	}

	resp, err := s.client.GetInfo(context.Background(), req)

	if err != nil {
		fmt.Printf("Cannot get the device info: %v\n", err)

		return nil, err
	}

	return resp.GetInfo(), nil
}

func (s *DeviceSvc) GetCapabilityInfo(deviceID uint32) (*device.CapabilityInfo, error) {
	req := &device.GetCapabilityInfoRequest{
		DeviceID: deviceID,
	}

	resp, err := s.client.GetCapabilityInfo(context.Background(), req)

	if err != nil {
		fmt.Printf("Cannot get the device capability info: %v\n", err)

		return nil, err
	}

	return resp.GetCapInfo(), nil
}

File: client/python/example/device/device.py
import grpc

import device_pb2_grpc
import device_pb2


class DeviceSvc:
  stub = None

  def __init__(self, channel): 
    try:
      self.stub = device_pb2_grpc.DeviceStub(channel)
    except grpc.RpcError as e:
      print(f'Cannot get the device stub: {e}')
      raise

  def getInfo(self, deviceID):
    try:
      response = self.stub.GetInfo(device_pb2.GetInfoRequest(deviceID=deviceID))
      return response.info
    except grpc.RpcError as e:
      print(f'Cannot get the device info: {e}')
      raise

  def getCapInfo(self, deviceID):
    try:
      response = self.stub.GetCapabilityInfo(device_pb2.GetCapabilityInfoRequest(deviceID=deviceID))
      return response.capInfo
    except grpc.RpcError as e:
      print(f'Cannot get the capability info: {e}')
      raise

